File: src/analysis/analyze-page-structure.mjs
#!/usr/bin/env node
// Deterministic page structure analyzer.
// Detects content modules (FAQ, table, list, video, form, image gallery),
// computes per-section word/sentence counts and depth scores,
// and produces cross-competitor module frequency analysis.
//
// Usage: node analyze-page-structure.mjs --pages-dir <pages/>
// Outputs JSON to stdout. Same inputs always produce byte-identical output.

import { readFileSync, readdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

// --- CLI parsing ---
const args = process.argv.slice(2);
function flag(name) {
  const idx = args.indexOf(name);
  if (idx === -1 || idx + 1 >= args.length) return undefined;
  return args[idx + 1];
}

const pagesDir = flag('--pages-dir');
const outputPath = flag('--output') || null;

if (pagesDir === undefined) {
  console.error('Usage: node analyze-page-structure.mjs --pages-dir <pages/>');
  process.exit(1);
}

// --- Load page files (sorted for determinism) ---
const pageFiles = readdirSync(pagesDir)
  .filter(f => f.endsWith('.json'))
  .sort();

if (pageFiles.length === 0) {
  const json = JSON.stringify({ competitors: [], cross_competitor: { common_modules: [], rare_modules: [], module_frequency: {}, avg_word_count: 0, avg_sections: 0 } }, null, 2);
  if (outputPath) {
    writeFileSync(outputPath, json);
  } else {
    console.log(json);
  }
  process.exit(0);
}

// --- Helpers ---

function countWords(text) {
  return text.split(/\s+/).filter(Boolean).length;
}

function countSentences(text) {
  // Split on sentence-ending punctuation followed by space or end-of-string
  const sentences = text.split(/[.?!]+/).filter(s => s.trim().length > 0);
  return sentences.length;
}

function hasNumbers(text) {
  return /\d/.test(text);
}

function hasLists(signals) {
  return (signals.ordered_lists > 0 || signals.unordered_lists > 0);
}

function computeDepthScore(sentenceCount) {
  if (sentenceCount <= 2) return 'shallow';
  if (sentenceCount <= 6) return 'basic';
  return 'detailed';
}

// FAQ heading patterns (German + English)
const FAQ_HEADING_RE = /\b(faq|fragen|haeufig|frequently\s+asked|h.ufig)\b/i;

// Block/error page heading patterns — conservative: only clear bot-wall and error pages
const BLOCK_HEADING_RE = /why have i been blocked|access denied|403 forbidden|please verify|checking your browser|just a moment|enable javascript and cookies|attention required/i;

const MIN_WORD_COUNT = 200;

// Returns a string reason if the page should be excluded, or null if it's acceptable.
function blockReason(mainText, headingTexts) {
  if (mainText.length === 0) return 'missing main_content_text';
  const wc = countWords(mainText);
  if (wc < MIN_WORD_COUNT) return `too few words (${wc} < ${MIN_WORD_COUNT})`;
  const blockedHeading = headingTexts.find(t => BLOCK_HEADING_RE.test(t));
  if (blockedHeading) return `block/error heading: "${blockedHeading}"`;
  return null;
}

function detectModules(signals, headingTexts) {
  const modules = [];

  // FAQ: heading text match or html_signals.faq_sections > 0
  const hasFaqHeading = headingTexts.some(t => FAQ_HEADING_RE.test(t));
  if (hasFaqHeading || signals.faq_sections > 0) {
    modules.push('faq');
  }

  if (signals.tables > 0) modules.push('table');
  if (signals.ordered_lists > 0 || signals.unordered_lists > 0) modules.push('list');
  if (signals.video_embeds > 0) modules.push('video');
  if (signals.images_in_content > 3) modules.push('image_gallery');
  if (signals.forms > 0) modules.push('form');

  // Sort for deterministic output
  modules.sort();
  return modules;
}

// Split main_content_text into sections based on heading text markers.
// The text between consecutive heading texts forms a section.
function splitSections(mainText, headings) {
  if (headings.length === 0) {
    // No headings: entire text is one section
    return [{ heading: '', level: 0, text: mainText }];
  }

  const sections = [];

  // Find positions of each heading in the text
  const positions = [];
  for (const h of headings) {
    const idx = mainText.indexOf(h.text);
    if (idx >= 0) {
      positions.push({ heading: h.text, level: h.level, pos: idx });
    }
  }

  // Sort by position in text (should already be in order, but ensure determinism)
  positions.sort((a, b) => a.pos - b.pos);

  // Intro section (text before first heading)
  if (positions.length > 0 && positions[0].pos > 0) {
    const introText = mainText.slice(0, positions[0].pos).trim();
    if (introText.length > 0) {
      sections.push({ heading: '', level: 0, text: introText });
    }
  }

  // Sections between headings
  for (let i = 0; i < positions.length; i++) {
    const start = positions[i].pos + positions[i].heading.length;
    const end = (i + 1 < positions.length) ? positions[i + 1].pos : mainText.length;
    const sectionText = mainText.slice(start, end).trim();
    sections.push({
      heading: positions[i].heading,
      level: positions[i].level,
      text: sectionText,
    });
  }

  return sections;
}

// --- Process each page ---
console.error(`Analyzing page structure for ${pageFiles.length} competitors...`);
const competitors = [];

for (const file of pageFiles) {
  const page = JSON.parse(readFileSync(join(pagesDir, file), 'utf-8'));
  const mainText = page.main_content_text || '';
  const headings = page.headings || [];
  const signals = page.html_signals || {
    faq_sections: 0, tables: 0, ordered_lists: 0,
    unordered_lists: 0, video_embeds: 0, forms: 0, images_in_content: 0,
  };

  const headingTexts = headings.map(h => h.text);
  const detectedModules = detectModules(signals, headingTexts);

  const rawSections = splitSections(mainText, headings);
  const sections = rawSections.map(s => {
    const wc = countWords(s.text);
    const sc = countSentences(s.text);
    return {
      heading: s.heading,
      level: s.level,
      word_count: wc,
      sentence_count: sc,
      has_numbers: hasNumbers(s.text),
      has_lists: hasLists(signals),
      depth_score: computeDepthScore(sc),
    };
  });

  // Extract domain from URL
  let domain = '';
  try {
    domain = new URL(page.url).hostname;
  } catch { /* skip */ }

  // Quality filter: skip blocked/error/thin pages
  const reason = blockReason(mainText, headingTexts);
  if (reason !== null) {
    process.stderr.write(`Skipping ${domain || file}: ${reason}\n`);
    continue;
  }

  competitors.push({
    url: page.url || '',
    domain,
    total_word_count: countWords(mainText),
    section_count: sections.length,
    detected_modules: detectedModules,
    sections,
  });
}

// --- Cross-competitor analysis ---
const totalCompetitors = competitors.length;

// Count module occurrences across competitors
const moduleCounts = {};
for (const comp of competitors) {
  for (const mod of comp.detected_modules) {
    moduleCounts[mod] = (moduleCounts[mod] || 0) + 1;
  }
}

// Sort module_frequency keys for deterministic output
const sortedModuleKeys = Object.keys(moduleCounts).sort();
const module_frequency = {};
for (const k of sortedModuleKeys) {
  module_frequency[k] = moduleCounts[k];
}

// Common: present in >= 70% of competitors
// Rare: present in <= 20% of competitors
const commonThreshold = totalCompetitors * 0.7;
const rareThreshold = totalCompetitors * 0.2;

const common_modules = sortedModuleKeys
  .filter(k => moduleCounts[k] >= commonThreshold)
  .sort();

const rare_modules = sortedModuleKeys
  .filter(k => moduleCounts[k] <= rareThreshold)
  .sort();

const totalWordCount = competitors.reduce((sum, c) => sum + c.total_word_count, 0);
const totalSections = competitors.reduce((sum, c) => sum + c.section_count, 0);

const avg_word_count = Math.round(totalWordCount / totalCompetitors);
const avg_sections = Math.round(totalSections / totalCompetitors);

const output = {
  competitors,
  cross_competitor: {
    common_modules,
    rare_modules,
    module_frequency,
    avg_word_count,
    avg_sections,
  },
};

const json = JSON.stringify(output, null, 2);
if (outputPath) {
  writeFileSync(outputPath, json);
} else {
  console.log(json);
}

File: seo_pipeline/analysis/analyze_page_structure.py
"""Deterministic page structure analyzer.

Detects content modules (FAQ, table, list, video, form, image gallery),
computes per-section word/sentence counts and depth scores,
and produces cross-competitor module frequency analysis.

Usage:
    python -m seo_pipeline.analysis.analyze_page_structure \
        --pages-dir <pages/> [--output path]
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from collections import Counter
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

from seo_pipeline.models.analysis import (
    CompetitorPageStructure,
    CrossCompetitorAnalysis,
    PageStructure,
    PageStructureSection,
)
from seo_pipeline.utils.math import js_round, normalize_number

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

MIN_WORD_COUNT = 200

FAQ_HEADING_RE = re.compile(
    r"\b(faq|fragen|haeufig|frequently\s+asked|h.ufig)\b", re.IGNORECASE
)

BLOCK_HEADING_RE = re.compile(
    r"why have i been blocked|access denied|403 forbidden|please verify"
    r"|checking your browser|just a moment|enable javascript and cookies"
    r"|attention required",
    re.IGNORECASE,
)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _count_words(text: str) -> int:
    """Count whitespace-separated non-empty tokens."""
    return len(text.split())


def _count_sentences(text: str) -> int:
    """Count sentences by splitting on sentence-ending punctuation.

    Mirrors the Node.js logic: split on [.?!]+ then count non-empty parts.
    """
    parts = re.split(r"[.?!]+", text)
    return sum(1 for s in parts if s.strip())


def _has_numbers(text: str) -> bool:
    """Return True if text contains at least one digit."""
    return bool(re.search(r"\d", text))


def _has_lists(signals: dict[str, Any]) -> bool:
    """Return True if the page has ordered or unordered lists."""
    return signals.get("ordered_lists", 0) > 0 or signals.get("unordered_lists", 0) > 0


def _compute_depth_score(sentence_count: int) -> str:
    """Classify section depth based on sentence count."""
    if sentence_count <= 2:
        return "shallow"
    if sentence_count <= 6:
        return "basic"
    return "detailed"


# ---------------------------------------------------------------------------
# Page quality filter
# ---------------------------------------------------------------------------


def _block_reason(
    main_text: str, heading_texts: list[str]
) -> str | None:
    """Return a reason string if the page should be excluded, else None."""
    if not main_text:
        return "missing main_content_text"
    wc = _count_words(main_text)
    if wc < MIN_WORD_COUNT:
        return f"too few words ({wc} < {MIN_WORD_COUNT})"
    for t in heading_texts:
        if BLOCK_HEADING_RE.search(t):
            return f'block/error heading: "{t}"'
    return None


# ---------------------------------------------------------------------------
# Module detection
# ---------------------------------------------------------------------------


def _detect_modules(
    signals: dict[str, Any], heading_texts: list[str]
) -> list[str]:
    """Detect content modules present on the page.

    Returns an alphabetically sorted list of module names.
    """
    modules: list[str] = []

    has_faq_heading = any(FAQ_HEADING_RE.search(t) for t in heading_texts)
    if has_faq_heading or signals.get("faq_sections", 0) > 0:
        modules.append("faq")

    if signals.get("tables", 0) > 0:
        modules.append("table")
    if signals.get("ordered_lists", 0) > 0 or signals.get("unordered_lists", 0) > 0:
        modules.append("list")
    if signals.get("video_embeds", 0) > 0:
        modules.append("video")
    if signals.get("images_in_content", 0) > 3:
        modules.append("image_gallery")
    if signals.get("forms", 0) > 0:
        modules.append("form")

    modules.sort()
    return modules


# ---------------------------------------------------------------------------
# Section splitting
# ---------------------------------------------------------------------------


def _split_sections(
    main_text: str, headings: list[dict[str, Any]]
) -> list[dict[str, Any]]:
    """Split main_content_text into sections based on heading positions."""
    if not headings:
        return [{"heading": "", "level": 0, "text": main_text}]

    positions: list[dict[str, Any]] = []
    for h in headings:
        idx = main_text.find(h["text"])
        if idx >= 0:
            positions.append(
                {"heading": h["text"], "level": h["level"], "pos": idx}
            )
    positions.sort(key=lambda p: p["pos"])

    sections: list[dict[str, Any]] = []

    # Intro before first heading
    if positions and positions[0]["pos"] > 0:
        intro = main_text[: positions[0]["pos"]].strip()
        if intro:
            sections.append({"heading": "", "level": 0, "text": intro})

    for i, pos in enumerate(positions):
        start = pos["pos"] + len(pos["heading"])
        end = positions[i + 1]["pos"] if i + 1 < len(positions) else len(main_text)
        sections.append(
            {
                "heading": pos["heading"],
                "level": pos["level"],
                "text": main_text[start:end].strip(),
            }
        )

    return sections


# ---------------------------------------------------------------------------
# Main analysis function
# ---------------------------------------------------------------------------


def analyze_page_structure(pages_dir: Path) -> PageStructure:
    """Analyze page structure from competitor pages.

    Args:
        pages_dir: Directory containing extracted page JSON files.

    Returns:
        PageStructure model with per-competitor structure and cross-competitor
        module frequency analysis.
    """
    page_files = sorted(p for p in pages_dir.iterdir() if p.suffix == ".json")

    empty_cross = CrossCompetitorAnalysis(
        common_modules=[],
        rare_modules=[],
        module_frequency={},
        avg_word_count=0,
        avg_sections=0,
    )

    if not page_files:
        return PageStructure(competitors=[], cross_competitor=empty_cross)

    print(
        f"Analyzing page structure for {len(page_files)} competitors...",
        file=sys.stderr,
    )

    competitors: list[CompetitorPageStructure] = []

    for f in page_files:
        raw = json.loads(f.read_text(encoding="utf-8"))
        main_text: str = raw.get("main_content_text", "")
        headings: list[dict[str, Any]] = raw.get("headings", [])
        signals: dict[str, Any] = raw.get("html_signals", {
            "faq_sections": 0,
            "tables": 0,
            "ordered_lists": 0,
            "unordered_lists": 0,
            "video_embeds": 0,
            "forms": 0,
            "images_in_content": 0,
        })

        heading_texts = [h["text"] for h in headings]
        detected_modules = _detect_modules(signals, heading_texts)

        raw_sections = _split_sections(main_text, headings)
        sections = [
            PageStructureSection(
                heading=s["heading"],
                level=s["level"],
                word_count=_count_words(s["text"]),
                sentence_count=_count_sentences(s["text"]),
                has_numbers=_has_numbers(s["text"]),
                has_lists=_has_lists(signals),
                depth_score=_compute_depth_score(_count_sentences(s["text"])),
            )
            for s in raw_sections
        ]

        # Extract domain from URL
        domain = ""
        try:
            domain = urlparse(raw.get("url", "")).hostname or ""
        except Exception:
            pass

        # Quality filter: skip blocked/error/thin pages
        reason = _block_reason(main_text, heading_texts)
        if reason is not None:
            print(
                f"Skipping {domain or f.name}: {reason}",
                file=sys.stderr,
            )
            continue

        competitors.append(
            CompetitorPageStructure(
                url=raw.get("url", ""),
                domain=domain,
                total_word_count=_count_words(main_text),
                section_count=len(sections),
                detected_modules=detected_modules,
                sections=sections,
            )
        )

    # --- Cross-competitor analysis ---
    total_competitors = len(competitors)

    if total_competitors == 0:
        return PageStructure(competitors=[], cross_competitor=empty_cross)

    module_counts: Counter[str] = Counter()
    for comp in competitors:
        for mod in comp.detected_modules:
            module_counts[mod] += 1

    # Sort keys alphabetically for deterministic output
    sorted_keys = sorted(module_counts.keys())
    module_frequency = {k: module_counts[k] for k in sorted_keys}

    common_threshold = total_competitors * 0.7
    rare_threshold = total_competitors * 0.2

    common_modules = sorted(
        k for k in sorted_keys if module_counts[k] >= common_threshold
    )
    rare_modules = sorted(
        k for k in sorted_keys if module_counts[k] <= rare_threshold
    )

    total_word_count = sum(c.total_word_count for c in competitors)
    total_sections = sum(c.section_count for c in competitors)

    avg_word_count = js_round(total_word_count / total_competitors)
    avg_sections = js_round(total_sections / total_competitors)

    cross_competitor = CrossCompetitorAnalysis(
        common_modules=common_modules,
        rare_modules=rare_modules,
        module_frequency=module_frequency,
        avg_word_count=avg_word_count,
        avg_sections=avg_sections,
    )

    return PageStructure(competitors=competitors, cross_competitor=cross_competitor)


# ---------------------------------------------------------------------------
# Serialization helper
# ---------------------------------------------------------------------------


def _normalize_tree(obj: Any) -> Any:
    """Recursively apply normalize_number to all numeric values."""
    if isinstance(obj, dict):
        return {k: _normalize_tree(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_normalize_tree(v) for v in obj]
    if isinstance(obj, float):
        return normalize_number(obj)
    return obj


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Analyze page structure from competitor pages."
    )
    parser.add_argument(
        "--pages-dir",
        required=True,
        type=Path,
        help="Directory with page JSON files.",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=None,
        help="Output file path (default: stdout).",
    )
    return parser


def main(argv: list[str] | None = None) -> None:
    """CLI entry point."""
    args = _build_parser().parse_args(argv)
    result = analyze_page_structure(args.pages_dir)
    output_dict = _normalize_tree(result.model_dump())
    output_json = json.dumps(output_dict, indent=2, ensure_ascii=False)

    if args.output:
        args.output.write_text(output_json, encoding="utf-8")
    else:
        sys.stdout.write(output_json + "\n")


if __name__ == "__main__":
    main()
